numbat/src/list.rs

```diff
@@ -0,0 +1,297 @@
+//! This module defines a custom kind of list used in the [`numbat::Value::List`].
+//! The specificities of this list are that:
+//! - It's based on a `VecDeque`, which means insertion at the beginning or the end is `O(1)`.
+//! - It's stored behind an `Arc`, which makes cloning the value very cheap.
+//! - It tries to reduce the number of allocations as much as possible, even when shared between multiple values.
+//!   If you are the only owner of the list or while you're not pushing elements in it in, never re-allocate.
+
+use std::{collections::VecDeque, fmt, sync::Arc};
+
+use crate::{value::Value, RuntimeError};
+
+/// Reference counted list / list view
+#[derive(Clone, Eq)]
+pub struct NumbatList<T> {
+    /// The original alloc shared between all values
+    alloc: Arc<VecDeque<T>>,
+    /// The indexes accessible to us. If `None` we own the whole allocation
+    view: Option<(usize, usize)>,
+}
+
+impl<T> Default for NumbatList<T> {
+    fn default() -> Self {
+        Self {
+            alloc: Default::default(),
+            view: Default::default(),
+        }
+    }
+}
+
+impl<T: fmt::Debug + Clone> fmt::Debug for NumbatList<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_list().entries(self.iter()).finish()
+    }
+}
+
+impl<T: PartialEq> PartialEq for NumbatList<T> {
+    fn eq(&self, other: &Self) -> bool {
+        // Best case scenario, the slice don't have the same len and we can early exit
+        if self.len() != other.len() {
+            return false;
+        }
+        // Second best case scenario, the slice comes from the same allocation
+        // and have the same view => they are equal
+        if Arc::as_ptr(&self.alloc) == Arc::as_ptr(&other.alloc) && self.view == other.view {
```

Comment:
Maybe use [`Arc::ptr_eq`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.ptr_eq) for these sort of checks?

```suggestion
        if Arc::ptr_eq(&self.alloc, &other.alloc) && self.view == other.view {
```